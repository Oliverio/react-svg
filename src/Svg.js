import React, { Component, forwardRef } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import './Svg.css';

export class Svg extends Component {
    static getDerivedStateFromProps(props, state) {
        return {
            contents: typeof props.svg === 'string' ? props.svg : state.contents,
        };
    }

    promise = undefined;
    state = { contents: '' };

    componentDidMount() {
        this.maybeWaitForSvg();
    }

    componentDidUpdate(prevProps) {
        if (this.props.svg !== prevProps.svg) {
            this.maybeWaitForSvg();
        }
    }

    componentWillUnmount() {
        this.resetWaitForSvg();
    }

    render() {
        const { svg, className, forwardedRef, ...rest } = this.props;
        const { contents } = this.state;
        const finalProps = {
            ...rest,
            className: classNames('svgWrapper', className),
        };

        return <i { ...finalProps } ref={ forwardedRef } dangerouslySetInnerHTML={ { __html: contents } } />;
    }

    resetWaitForSvg() {
        this.promise = undefined;
    }

    async maybeWaitForSvg() {
        const { svg } = this.props;

        this.resetWaitForSvg();

        if (typeof svg !== 'object') {
            return;
        }

        const promise = this.promise = svg;
        const result = await this.promise.catch(() => ({ default: '' }));

        /* istanbul ignore next */
        if (promise === this.promise) {
            this.setState({ contents: result.default });
        }
    }
}

Svg.propTypes = {
    svg: PropTypes.oneOfType([PropTypes.string, PropTypes.shape({ then: PropTypes.func.isRequired })]).isRequired,
    className: PropTypes.string,
    forwardedRef: PropTypes.object,
};

export default forwardRef((props, ref) => <Svg forwardedRef={ ref } { ...props } />);
