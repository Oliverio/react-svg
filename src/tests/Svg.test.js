import React from 'react';
import { render } from '@testing-library/react';
import Svg from '../Svg';

const svg = `<svg height="100" width="100"> 
                <circle cx="50" cy="50" r="40" stroke="black" stroke-width="3"/>
            </svg>`;

const defaultProps = {
    svg,
};

const renderWithProps = (props = {}) => render(<Svg { ...defaultProps } { ...props } />);

describe('Svg', () => {
    it('should render correctly', () => {
        const { asFragment } = renderWithProps();

        expect(asFragment()).toMatchSnapshot();
    });

    it('should handle correctly when svg is a promise', () => {
        const spy = Promise.resolve({ default: svg });
        const { asFragment } = renderWithProps({ svg: spy });

        expect(asFragment()).toMatchSnapshot();
    });

    it('should handle correctly when svg is a promise that fails', () => {
        const spy = Promise.reject(new Error('foo'));
        const { asFragment } = renderWithProps({ svg: spy });

        expect(asFragment()).toMatchSnapshot();
    });

    it('should render correctly when receiving new svg prop', () => {
        const { asFragment, rerender } = renderWithProps();
        const spy = Promise.resolve({ default: svg });

        rerender(<Svg svg={ spy } />);

        expect(asFragment()).toMatchSnapshot();
    });

    it('should render correctly when receiving new svg promise', () => {
        jest.useFakeTimers();

        const spy = new Promise((resolve) => {
            setTimeout(() => { resolve({ default: svg }); }, 500);
        });
        const spy2 = new Promise((resolve) => {
            setTimeout(() => { resolve({ default: svg }); }, 500);
        });

        const { asFragment, rerender } = renderWithProps({ svg: spy });

        rerender(<Svg svg={ spy2 } />);

        jest.runAllTimers();

        expect(asFragment()).toMatchSnapshot();
    });
});
