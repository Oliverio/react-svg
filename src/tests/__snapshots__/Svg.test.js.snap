// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Svg should handle correctly when svg is a promise 1`] = `<DocumentFragment />`;

exports[`Svg should handle correctly when svg is a promise that fails 1`] = `<DocumentFragment />`;

exports[`Svg should render correctly 1`] = `
<DocumentFragment>
  <svg
    class="react-svg"
    height="100"
    width="100"
  >
     
                
    <circle
      cx="50"
      cy="50"
      r="40"
      stroke="black"
      stroke-width="3"
    />
    
            
  </svg>
</DocumentFragment>
`;

exports[`Svg should render correctly when receiving new svg promise 1`] = `<DocumentFragment />`;

exports[`Svg should render correctly when receiving new svg prop 1`] = `
<DocumentFragment>
  <svg
    class="react-svg"
    height="100"
    width="100"
  >
     
                
    <circle
      cx="50"
      cy="50"
      r="40"
      stroke="black"
      stroke-width="3"
    />
    
            
  </svg>
</DocumentFragment>
`;
